packages/@aws-cdk/core/lib/private/synthesis.ts

```diff
@@ -49,7 +55,97 @@ export function synthesize(root: IConstruct, options: SynthesisOptions = { }): c
   // stacks to add themselves to the synthesized cloud assembly.
   synthesizeTree(root, builder, options.validateOnSynthesis);
 
-  return builder.buildAssembly();
+  const assembly = builder.buildAssembly();
+
+  invokeValidationPlugins(root, builder.outdir, options.validationReportFormat);
+
+  return assembly;
+}
+
+class LazyHash {
+  private hash?: string;
+
+  constructor(private readonly outdir: string) {}
+
+  get value(): string {
+    if (!this.hash) {
+      this.hash = computeChecksumOfFolder(this.outdir);
+    }
+    return this.hash;
+  }
+}
+
+function invokeValidationPlugins(root: IConstruct, outdir: string, validationReportFormat?: ValidationReportFormat) {
+  const lazyHash = new LazyHash(outdir);
+
+  const templatePathsByPlugin: Map<IValidationPlugin, string[]> = new Map();
+  visit(root, 'post', construct => {
+    if (Stage.isStage(construct)) {
+      for (const plugin of construct.validationPlugins) {
+        if (!templatePathsByPlugin.has(plugin)) {
+          templatePathsByPlugin.set(plugin, []);
+        }
+        templatePathsByPlugin.get(plugin)!.push(...templatePaths(construct.outdir));
+      }
+    }
+  });
+
+  const reports: ValidationPluginReport[] = [];
+  for (const [plugin, paths] of templatePathsByPlugin.entries()) {
+    const originalHash = lazyHash.value;
+
+    try {
+      const report = plugin.validate({ templatePaths: paths });
+      reports.push(report);
+    } catch (e) {
```

Comment:
This error should be incorporated in the report somehow

Reply:
You're right. Done.